usr/src/man/man1/ckstr.1
'\" te
.\"  Copyright 1989 AT&T  Copyright (c) 2001, Sun Microsystems, Inc.  All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH CKSTR 1 "Sep 14, 1992"
.SH NAME
ckstr, errstr, helpstr, valstr \- display a prompt; verify and return a string
answer
.SH SYNOPSIS
.LP
.nf
\fBckstr\fR [\fB-Q\fR] [\fB-W\fR \fIwidth\fR] [ [\fB-r\fR \fIregexp\fR] [...]] [\fB-l\fR \fIlength\fR]
     [\fB-d\fR \fIdefault\fR] [\fB-h\fR \fIhelp\fR] [\fB-e\fR \fIerror\fR] [\fB-p\fR \fIprompt\fR]
     [\fB-k\fR \fIpid\fR [\fB- s\fR \fIsignal\fR]]
.fi

.LP
.nf
\fB/usr/sadm/bin/errstr\fR [\fB-W\fR \fIwidth\fR] [\fB-e\fR \fIerror\fR] [\fB-l\fR \fIlength\fR]
     [ [\fB-r\fR \fIregexp\fR] [...]]
.fi

.LP
.nf
\fB/usr/sadm/bin/helpstr\fR [\fB-W\fR \fIwidth\fR] [\fB-h\fR \fIhelp\fR] [\fB-l\fR \fIlength\fR]
     [ [\fB-r\fR \fIregexp\fR] [...]]
.fi

.LP
.nf
\fB/usr/sadm/bin/valstr\fR [\fB-l\fR \fIlength\fR] [ [\fB-r\fR \fIregexp\fR] [...]] \fIinput\fR
.fi

.SH DESCRIPTION
.sp
.LP
The \fBckstr\fR utility prompts a user and validates the response. It defines,
among other things, a prompt message whose response should be a string, text
for help and error messages, and a default value (which are returned if the
user responds with a RETURN).
.sp
.LP
The answer returned from this command must match the defined regular expression
and be no longer than the length specified. If no regular expression is given,
valid input must be a string with a length less than or equal to the length
defined with no internal, leading or trailing white space. If no length is
defined, the length is not checked.
.sp
.LP
All messages are limited in length to 79 characters and are formatted
automatically. Tabs and newlines are removed after a single white space
character in a message definition, but spaces are not removed. When a tilde is
placed at the beginning or end of a message definition, the default text will
be inserted at that point, allowing both custom text and the default text to be
displayed.
.sp
.LP
If the prompt, help or error message is not defined, the default message (as
defined under EXAMPLES) is displayed.
.sp
.LP
Three visual tool modules are linked to the \fBckstr\fR command. They are
\fBerrstr\fR (which formats and displays an error message on the standard
output), \fBhelpstr\fR (which formats and displays a help message on the
standard output), and \fBvalstr\fR (which validates a response).
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-d\fR \fIdefault\fR\fR
.ad
.RS 14n
Defines the default value as \fIdefault\fR. The default is not validated and so
does not have to meet any criteria.
.RE

.sp
.ne 2
.na
\fB\fB-e\fR \fIerror\fR\fR
.ad
.RS 14n
Defines the error message as \fI error\fR.
.RE

.sp
.ne 2
.na
\fB\fB-h\fR \fIhelp\fR\fR
.ad
.RS 14n
Defines the help message as \fI help\fR.
.RE

.sp
.ne 2
.na
\fB\fB-k\fR \fIpid\fR\fR
.ad
.RS 14n
Specifies that process \fBID\fR \fIpid\fR is to be sent a signal if the user
chooses to quit.
.RE

.sp
.ne 2
.na
\fB\fB-l\fR \fIlength\fR\fR
.ad
.RS 14n
Specifies the maximum length of the input.
.RE

.sp
.ne 2
.na
\fB\fB-p\fR \fIprompt\fR\fR
.ad
.RS 14n
Defines the prompt message as \fIprompt\fR.
.RE

.sp
.ne 2
.na
\fB\fB-Q\fR\fR
.ad
.RS 14n
Specifies that \fBquit\fR will not be allowed as a valid response.
.RE

.sp
.ne 2
.na
\fB\fB-r\fR \fIregexp\fR\fR
.ad
.RS 14n
Specifies a regular expression, \fI regexp\fR, against which the input should
be validated. May include white space. If multiple expressions are defined, the
answer need match only one of them.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR \fIsignal\fR\fR
.ad
.RS 14n
Specifies that the process \fBID\fR \fIpid\fR defined with the \fB-k\fR option
is to be sent signal \fIsignal\fR when quit is chosen. If no signal is
specified, \fBSIGTERM\fR is used.
.RE

.sp
.ne 2
.na
\fB\fB-W\fR \fIwidth\fR\fR
.ad
.RS 14n
Specifies that prompt, help and error messages will be formatted to a line
length of \fIwidth\fR.
.RE

.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.na
\fB\fIinput\fR\fR
.ad
.RS 9n
Input to be verified against format length and/or regular expression criteria.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRDefault prompt
.sp
.LP
The default prompt for \fBckstr\fR is:

.sp
.in +2
.nf
example% \fBckstr\fR
Enter an appropriate value [?,q]:
.fi
.in -2
.sp

.LP
\fBExample 2 \fRDefault error message
.sp
.LP
The default error message is dependent upon the type of validation involved.
The user will be told either that the length or the pattern matching failed.
The default error message is:

.sp
.in +2
.nf
example% \fB/usr/sadm/bin/errstr\fR
ERROR: Please enter a string which contains no embedded,
leading or trailing spaces or tabs.
.fi
.in -2
.sp

.LP
\fBExample 3 \fRDefault help message
.sp
.LP
The default help message is also dependent upon the type of validation
involved.  If a regular expression has been defined, the message is:

.sp
.in +2
.nf
example% \fB/usr/sadm/bin/helpstr \fR\fB-r\fR\fB regexp\fR
Please enter a string which matches the following pattern:
regexp
.fi
.in -2
.sp

.sp
.LP
Other messages define the length requirement and the definition of a string.

.LP
\fBExample 4 \fRUsing the quit option
.sp
.LP
When the quit option is chosen (and allowed), \fBq\fR is returned along with
the return code \fB3\fR. Quit input gets a trailing newline.

.LP
\fBExample 5 \fRUsing the valstr module
.sp
.LP
The \fBvalstr\fR module will produce a usage message on stderr. It returns
\fB0\fR for success and non-zero for failure.

.sp
.in +2
.nf
example% \fB/usr/sadm/bin/valstr\fR
usage: valstr [\fB-l\fR length] [[\fB-r\fR regexp] [\|.\|.\|.\|]] input
.fi
.in -2
.sp

.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 5n
Successful execution.
.RE

.sp
.ne 2
.na
\fB\fB1\fR\fR
.ad
.RS 5n
\fBEOF\fR on input, or negative width on  \fB-W\fR option,  or usage error.
.RE

.sp
.ne 2
.na
\fB\fB2\fR\fR
.ad
.RS 5n
Invalid regular expression.
.RE

.sp
.ne 2
.na
\fB\fB3\fR\fR
.ad
.RS 5n
User termination (quit).
.RE

.SH SEE ALSO
.sp
.LP
.BR signal.h (3HEAD),
.BR attributes (7)


usr/src/lib/libadm/common/ckstr.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T */
/*	  All Rights Reserved  	*/


/*
 * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#include <stdio.h>
#include <string.h>
#include <limits.h>
#include <sys/types.h>
#include "libadm.h"

/*
 * This file is the only one anywhere to need these functions,
 * so we declare them here, not in libadm.h
 */
extern char *__compile(char *, char *, const char *, int);
extern int __step(const char *, const char *);

#define	ESIZE	1024

#define	ERRMSG0 "Input is required."
#define	ERRMSG1	"Please enter a string containing no more than %d characters."
#define	ERRMSG2	\
	"Pattern matching has failed."
#define	ERRMSG3 \
	"Please enter a string which contains no embedded, \
	leading or trailing spaces or tabs."

#define	HLPMSG0	"Please enter a string"
#define	HLPMSG1 "Please enter a string containing no more than %d characters"
#define	HLPMSG2 "matches one of the following patterns:"
#define	HLPMSG3 "matches the following pattern:"
#define	HLPMSG4 "contains no embedded, leading or trailing spaces or tabs."

static char	*errstr;

static char *
sethlp(char *msg, char *regexp[], int length)
{
	int	i;

	if (length)
		(void) sprintf(msg, HLPMSG1, length);
	else
		(void) strcpy(msg, HLPMSG0);

	(void) strcat(msg, length ? " and " : " which ");

	if (regexp && regexp[0]) {
		(void) strcat(msg, regexp[1] ? HLPMSG2 : HLPMSG3);
		for (i = 0; regexp[i]; i++) {
			(void) strcat(msg, "\\n\\t");
			(void) strcat(msg, regexp[i]);
		}
	} else
		(void) strcat(msg, HLPMSG4);
	return (msg);
}

int
ckstr_val(char *regexp[], int length, char *input)
{
	char	expbuf[ESIZE];
	int	i, valid;

	valid = 1;
	if (length && (strlen(input) > (size_t)length)) {
		errstr = ERRMSG1;
		return (1);
	}
	if (regexp && regexp[0]) {
		valid = 0;
		for (i = 0; !valid && regexp[i]; ++i) {
			if (!__compile(regexp[i], expbuf, &expbuf[ESIZE], '\0'))
				return (2);
			valid = __step(input, expbuf);
		}
		if (!valid)
			errstr = ERRMSG2;
	} else if (strpbrk(input, " \t")) {
		errstr = ERRMSG3;
		valid = 0;
	}
	return (valid == 0);
}

void
ckstr_err(char *regexp[], int length, char *error, char *input)
{
	char	*defhlp;
	char	temp[1024];

	if (input) {
		if (ckstr_val(regexp, length, input)) {
			/* LINTED E_SEC_PRINTF_VAR_FMT */
			(void) snprintf(temp, sizeof (temp), errstr, length);
			puterror(stdout, temp, error);
			return;
		}
	}

	defhlp = sethlp(temp, regexp, length);
	puterror(stdout, defhlp, error);
}

void
ckstr_hlp(char *regexp[], int length, char *help)
{
	char	*defhlp;
	char	hlpbuf[1024];

	defhlp = sethlp(hlpbuf, regexp, length);
	puthelp(stdout, defhlp, help);
}

int
ckstr(char *strval, char *regexp[], int length, char *defstr, char *error,
	char *help, char *prompt)
{
	int	n;
	char	*defhlp;
	char	input[MAX_INPUT],
		hlpbuf[1024],
		errbuf[1024];

	defhlp = NULL;
	if (!prompt)
		prompt = "Enter an appropriate value";

start:
	putprmpt(stderr, prompt, NULL, defstr);
	if (getinput(input))
		return (1);

	n = (int)strlen(input);
	if (n == 0) {
		if (defstr) {
			(void) strcpy(strval, defstr);
			return (0);
		}
		puterror(stderr, ERRMSG0, error);
		goto start;
	}
	if (strcmp(input, "?") == 0) {
		if (defhlp == NULL)
			defhlp = sethlp(hlpbuf, regexp, length);
		puthelp(stderr, defhlp, help);
		goto start;
	}
	if (ckquit && (strcmp(input, "q") == 0)) {
		(void) strcpy(strval, input);
		return (3);
	}
	if (ckstr_val(regexp, length, input)) {
		/* LINTED E_SEC_PRINTF_VAR_FMT */
		(void) snprintf(errbuf, sizeof (errbuf), errstr, length);
		puterror(stderr, errbuf, error);
		goto start;
	}
	(void) strcpy(strval, input);
	return (0);
}
